### ISSUE TITLE
k8s support Deployment(Config) rollouts and rollbacks

### ISSUE THREAD

**Issue by Akasurde** [ansible-collections/community.kubernetes#138]
_From @djzager on Feb 11, 2019 20:58_

##### SUMMARY

https://kubernetes.io/docs/concepts/workloads/controllers/deployment/

In the same way that the [`k8s_scale` module](https://docs.ansible.com/ansible/latest/modules/k8s_scale_module.html#k8s-scale-module) allows you to scale up|down Kubernetes objects that support the scale subresource, it would be great to have a supported method for rolling out and rolling back Deployments (and DeploymentConfigs).

A specific use case that I have in mind would be to tie the configuration of the application (stored in a ConfigMap) with the version of the Deployment such that when the ConfigMap is updated (instead of deleting all of the active pods) we roll out a new version of the application.

##### ISSUE TYPE
- Feature Idea

##### COMPONENT NAME

Maybe one for `k8s_rollout` and another for `k8s_rollback`?

##### ADDITIONAL INFORMATION

```yaml
handlers:
  - name: upgrade myapp
    k8s_rollout:
      api_version: v1
      kind: Deployment
      name: myapp
      namespace: myapp-namespace

tasks:
  # Set the configmap, upgrade when changed
  - name: set configmap
    k8s:
      state: present
      definition:
        kind: ConfigMap
        apiVersion: v1
        metadata:
          name: myapp-config
          namespace: myapp-namespace
        data:
          foo: bar
    notify: upgrade myapp
```

_Copied from original issue: ansible/ansible#52046_

**Comment by Akasurde** [comment]
_From @djzager on Feb 13, 2019 15:55_

As far as the ConfigMap use case is concerned, I believe you could simply use the `append_hash` feature coming to the [`k8s` module in 2.8`](https://docs.ansible.com/ansible/devel/modules/k8s_module.html) to version your ConfigMap. In that case, an update to the ConfigMap would give it a new name and require a change also to the Deployment(Config) with a subsequent rollout.

**Comment by Akasurde** [comment]
_From @sivel on Feb 19, 2019 20:19_

!component =lib/ansible/modules/clustering/k8s/k8s.py

**Comment by Akasurde** [comment]
_From @willthames on Feb 19, 2019 23:42_

Further to what @djzager said, there's a [blog post](https://willthames.github.io/2019/01/28/immutable-kubernetes-configuration-with-ansible.html) and a [conference talk](https://www.ansible.com/managing-kubernetes-is-easy-with-ansible) on using immutable configmaps with ansible

**Comment by Akasurde** [comment]
_From @fabianvf on Oct 25, 2019 17:38_

It seems like rollouts are best handled with the immutable configmaps/secrets, but rollback would still be nice to have.

**Comment by tima** [comment]
@djzager In addition to the suggestions about rollouts, a `k8s_rollback` module was added by #26 and was released as part of 1.1. Is that sufficient that this one can be closed.
